keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1257 @@
+have container-specific QoS classes (as a possible future enhancement of K8s).
+
+Communicating Pod QoS class via class resources would advocate moving class
+resources up to `ContainerConfig`.
+
+Making this change, it would also be possible to separate `oom_score_adj` from
+the pod qos class in the future.  The runtime could provide a set of OOM
+classes, making it possible for the user to specify a burstable pod with low
+oom priority (low chance of being killed).
+
+#### Default class
+
+A mechanism for indicating that the (runtime) default class should be used. The
+default class would/should be a node/runtime specific attribute. How should
+this be specified in the CRI protocol/`cri-api` and Pod spec?
+
+### Future work
+
+This section sheds light on the end goal of this work in order to better
+evaluate this KEP in a broader context. What a fully working solution would
+consists of and what the (next) steps to accomplish that would be. These topics
+are currently out of the scope of this KEP and were listed under
+[Non-goals](#non-goals).
+
+#### Pod Spec
+
+Replace pod annotations with proper user interface via the Pod spec. Below, one
+possible option is presented.
+
+Introduce a new field (e.g. class) into ResourceRequirements of Container.
+
+```diff
+// ResourceRequirements describes the compute resource requirements.
+type ResourceRequirements struct {
+     // Limits describes the maximum amount of compute resources allowed.
+     Limits ResourceList `json:"limits,omitempty"
+     // Requests describes the minimum amount of compute resources required.
+     Requests ResourceList `json:"requests,omitempty"
++    // Classes specifies the class resources that the container should be assigned
++    Classes map[ClassResourceName]string
+}
+
++// ClassResourceName is the name of a class-based resource.
++type ClassResourceName string
+```
+
+Also, we add a `Resources` field to the `PodSpec`. We will re-use the existing
+`ResourceRequirements` type but Limits and Requests must be left empty. Classes
+may be set and they represent the Pod-level assignment of class resources,
+comparable to the PodClassResources message in PodSandboxConfig in the CRI API.
+
+```diff
+ type PodSpec struct {
+@@ -224,4 +224,8 @@ type PodSpec struct {
+     // Default to false.
+     // +optional
+     SetHostnameAsFQDN *bool `json:"setHostnameAsFQDN,omitempty" protobuf:"varint,35,opt,name=setHostnameAsFQDN"`
++    // Pod-level resources. Currently, requests and limits are not allowed
++    // to be specified for pods.
++    // +optional
++    Resources ResourceRequirements
+ }
+```
+
+In practice, the class resource information will be directly used in the CRI
+ContainerConfig (e.g.  CreateContainerRequest message). At this point, without
+resource discovery or access control kubelet does not do any validity checking
+of the values. Invalid class assignments will cause an error in the container
+runtime.
```

Comment:
in admission controller or on pod creation? Will it get into the state similar to ImagePullBackoff or it will infinitely retry?

This question may look like implementaiton detail, but what I really want to understand is the mutability of resource classes

Reply:
It would result in CreateContainerError with infinitely retry (if I understand kubelet's behavior correctly).

Reply:
@SergeyKanzhelev closing this as outdated. The new proposal contains kubelet admission checks already in the first implementation phase